src/service-worker/service-worker.ts

```diff
@@ -1,37 +1,57 @@
-import {Message, FileRecord} from '../shared/types.js';
-import {endWithSlash} from '../shared/util.js';
-import {expose} from 'comlink';
-import {MESSAGE_TYPES} from '../shared/constants.js';
-
-const swScope = (self as unknown) as ServiceWorkerGlobalScope;
+/**
+ * @license
+ * Copyright (c) 2019 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {
+  HANDSHAKE_RECEIVED,
+  ESTABLISH_HANDSHAKE,
+  ServiceWorkerAPI,
+  FileAPI,
+} from '../shared/worker-api.js';
+import { endWithSlash } from '../shared/util.js';
+import { expose } from 'comlink';
+
+declare var self: ServiceWorkerGlobalScope;
+
+/**
+ * API exposed to the UI thread via Comlink. The static methods on this class
+ * become instance methods on SwControllerAPI.
+ */
+const workerAPI: ServiceWorkerAPI = {
+  setFileAPI(fileAPI: FileAPI, sessionID: string) {
+    fileAPIs.set(sessionID, fileAPI);
+  },
+};
 
-const recieveMessageChannelHandshake = (e: MessageEvent) => {
-  const ports = e.ports;
-  if (ports && ports[0]) {
-    const port = ports[0];
-    port.start();
-    const handshakeReceivedMessage: Message = {
-      type: MESSAGE_TYPES.HANDSHAKE_RECEIVED,
-    };
-    expose(SwController, port);
-    port.postMessage(handshakeReceivedMessage);
+const fileAPIs = new Map<string, FileAPI>();
+const getFile = async (e: FetchEvent, path: string, sessionId: string) => {
+  const fileAPI = fileAPIs.get(sessionId);
+  if (fileAPI) {
+    const file = await fileAPI.getFile(path);
```

Comment:
How is performance on this? I remember creating the file record cache simply because you said it was taking too long to reload fetching from the client every time

Reply:
Hmm... I don't remember that, maybe it had to do with compilation? If we were going to do compilation in the service worker, we'd need all the files sent over. But I put compilation in a separate worker so that it wouldn't tie up the service worker thread and it could fetch other files for other projects in parallel. The perf here is fine, copying strings is fast.